### PR TITLE
98249 Implement draft appointment controller method (WIP)

### PR DESCRIPTION
**Pull request by randomsync**
**Note**: Delete the description statements, complete each step. **None are optional**, but can be justified as to why they cannot be completed as written. Provide known gaps to testing that may raise the risk of merging to production.

## Summary

- *This work is behind a feature toggle (flipper): YES/NO*
- *(Summarize the changes that have been made to the platform)*
- *(If bug, how to reproduce)*
- *(What is the solution, why is this the solution?)*
- *(Which team do you work for, does your team own the maintenance of this component?)*
- *(If introducing a flipper, what is the success criteria being targeted?)*

## Related issue(s)

- *Link to ticket created in va.gov-team repo OR screenshot of Jira ticket if your team uses Jira*
- *Link to previous change of the code/bug (if applicable)*
- *Link to epic if not included in ticket*

## Testing done

- [ ] *New code is covered by unit tests*
- *Describe what the old behavior was prior to the change*
- *Describe the steps required to verify your changes are working as expected. Exclusively stating 'Specs run' is NOT acceptable as appropriate testing*
- *If this work is behind a flipper:*
  - *Tests need to be written for both the flipper on and flipper off scenarios. [Docs](https://depo-platform-documentation.scrollhelp.site/developer-docs/feature-toggles-guide#Featuretogglesguide-Backendexample).*
  - *What is the testing plan for rolling out the feature?*

## Screenshots
_Note: Optional_

## What areas of the site does it impact?
*(Describe what parts of the site are impacted and*if*code touched other areas)*

## Acceptance criteria

- [ ]  I fixed|updated|added unit tests and integration tests for each feature (if applicable).
- [ ]  No error nor warning in the console.
- [ ]  Events are being sent to the appropriate logging solution
- [ ]  Documentation has been updated (link to documentation)
- [ ]  No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs
- [ ]  Feature/bug has a monitor built into Datadog (if applicable)
- [ ]  If app impacted requires authentication, did you login to a local build and verify all authenticated routes work as expected
- [ ]  I added a screenshot of the developed feature

## Requested Feedback

(OPTIONAL)_What should the reviewers know in addition to the above. Is there anything specific you wish the reviewer to assist with. Do you have any concerns with this PR, why?_

### CODE REVIEW

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 1]
```diff
@@ -62,6 +62,25 @@ def create
...
+
+        response_data = OpenStruct.new(
+          id: draft_appointment.id,
+          provider: provider,
+          slots: fetch_provider_slots,
+          drive_time: fetch_drive_times(provider)
+        )
+
+        serialized = Eps::DraftAppointmentSerializer.new(response_data)
+        render json: serialized, status: :created
```
Comment (kanchanasuriya) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:81]: What if one of the API call fails? Can we return `:multistatus` as the current `AppointmentsController` endpoint?  Please ignore, if this is planned in future PR.

File: modules/vaos/app/serializers/eps/draft_appointment_serializer.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+module Eps
```
Comment (kanchanasuriya) [inline modules/vaos/app/serializers/eps/draft_appointment_serializer.rb:3]: Assuming this serializer change is for just adding rspecs & final serializer changes come from [PR](https://github.com/department-of-veterans-affairs/vets-api/pull/20337)

Comment (randomsync) [reply modules/vaos/app/serializers/eps/draft_appointment_serializer.rb:3]: yes, that's correct

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb (hunk 1 above)
Comment (randomsync) [reply modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:81]: :multistatus doesn't apply here. This is 1 resource that is returned, not an array of resources.

**Review by kanchanasuriya — APPROVED** [review]
LGTM

**Comment by stevenjcumming** [comment]
@randomsync there's a failing test. Possibly a flakey test, please re-run the Code Check - Test until it passes.

File: modules/vaos/app/controllers/vaos/v2/appointments_controller.rb [hunk 3]
```diff
@@ -62,6 +62,25 @@ def create
...
       end
 
+      def create_draft
+        # TODO: validate referral_id from the cache from prior referrals response,
+        # TODO: validate that the referral doesn't already have a confirmed appointment #
+        # TODO: cache provider_id, appointment_type_id, end_date from prior referrals response and use here
+        draft_appointment = eps_appointment_service.create_draft_appointment(referral_id: draft_params[:referral_id])
+
+        provider = eps_provider_service.get_provider_service(provider_id: draft_params[:provider_id])
+
```
Comment (stevenjcumming) [inline modules/vaos/app/controllers/vaos/v2/appointments_controller.rb:72]: You could also make these private methods (with or without memoization as needed)

```ruby
private 

def draft_appointment
  eps_appointment_service.create_draft_appointment(referral_id: draft_params[:referral_id])
end

def provider 
  eps_provider_service.get_provider_service(provider_id: draft_params[:provider_id])
end
```

**Review by stevenjcumming — APPROVED** [review]

File: modules/vaos/spec/requests/vaos/v2/appointments_spec.rb [hunk 4]
```diff
@@ -880,4 +880,148 @@ def stub_clinics
...
+              'attributes' => {
+                'provider' => provider_response,
+                'slots' => slots_response['slots'],
+                'drivetime' => drive_times_response
+              }
+            }
+          }
+        end
+
+        it 'returns a successful response when all calls succeed' do
```
Comment (lee-delarm6) [inline modules/vaos/spec/requests/vaos/v2/appointments_spec.rb:1008]: Maybe more tests for when one of the calls fails or there's an empty param, etc.? I saw them as TODOs towards the top

**Review by lee-delarm6 — CHANGES_REQUESTED** [review]
Looks good, but maybe more tests for nil/empty objects or params.

Will those TODOs at the top be covered in this PR or another one?

File: modules/vaos/spec/requests/vaos/v2/appointments_spec.rb (hunk 4 above)
Comment (randomsync) [reply modules/vaos/spec/requests/vaos/v2/appointments_spec.rb:1008]: since this PR was already too large, I created a separate ticket for that and and it'll be worked on in this sprint: https://github.com/department-of-veterans-affairs/va.gov-team/issues/101428

Comment (randomsync) [reply modules/vaos/spec/requests/vaos/v2/appointments_spec.rb:1008]: @lee-delarm6 do you think we shouldn't merge this unless more tests are added?

**Review by devin-mccurdy — APPROVED** [review]
LGTM

**Review by lee-delarm6 — APPROVED** [review]
More tests will be added in another PR Gaurav has said, so good to go

**Review by stevenjcumming — APPROVED** [review]
